cluster-autoscaler/cloudprovider/hetzner/hetzner_cloud_provider.go

```diff
@@ -0,0 +1,242 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hetzner
+
+import (
+	"fmt"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
+	"k8s.io/autoscaler/cluster-autoscaler/config"
+	"k8s.io/autoscaler/cluster-autoscaler/utils/errors"
+	"k8s.io/klog/v2"
+	"regexp"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+)
+
+var _ cloudprovider.CloudProvider = (*HetznerCloudProvider)(nil)
+
+const (
+	// GPULabel is the label added to nodes with GPU resource.
+	GPULabel               = hcloudLabelNamespace + "/gpu-node"
+	providerIDPrefix       = "hcloud://"
+	nodeGroupLabel         = hcloudLabelNamespace + "/node-group"
+	hcloudLabelNamespace   = "hcloud"
+	drainingNodePoolId     = "draining-node-pool"
+	serverCreateTimeout    = 1 * time.Minute
+	serverRegisterTimeout  = 10 * time.Minute
+	defaultPodAmountsLimit = 110
+)
+
+// HetznerCloudProvider implements CloudProvider interface.
+type HetznerCloudProvider struct {
+	manager         *hetznerManager
+	resourceLimiter *cloudprovider.ResourceLimiter
+}
+
+// Name returns name of the cloud provider.
+func (d *HetznerCloudProvider) Name() string {
+	return cloudprovider.HetznerProviderName
+}
+
+// NodeGroups returns all node groups configured for this cloud provider.
+func (d *HetznerCloudProvider) NodeGroups() []cloudprovider.NodeGroup {
+	groups := make([]cloudprovider.NodeGroup, 0)
+	for i := range d.manager.nodeGroups {
+		groups = append(groups, d.manager.nodeGroups[i])
+	}
+	return groups
+}
+
+// NodeGroupForNode returns the node group for the given node, nil if the node
+// should not be processed by cluster autoscaler, or non-nil error if such
+// occurred. Must be implemented.
+func (d *HetznerCloudProvider) NodeGroupForNode(node *apiv1.Node) (cloudprovider.NodeGroup, error) {
+	server, err := d.manager.serverForNode(node)
+	if err != nil {
+		return nil, fmt.Errorf("failed to check if server %s exists error: %v", node.Spec.ProviderID, err)
+	}
+
+	var groupId string
+	if server == nil {
+		klog.V(3).Infof("failed to find hcloud server for node %s", node.Name)
+		nodeGroupId, exists := node.Labels[nodeGroupLabel]
+		if !exists {
+			return nil, nil
+		}
+		groupId = nodeGroupId
+	} else {
+		serverGroupId, exists := server.Labels[nodeGroupLabel]
+		groupId = serverGroupId
+		if !exists {
+			return nil, nil
+		}
+	}
+
+	group, exists := d.manager.nodeGroups[groupId]
+	if !exists {
+		return nil, nil
+	}
+
+	return group, nil
+}
+
+// Pricing returns pricing model for this cloud provider or error if not
+// available. Implementation optional.
+func (d *HetznerCloudProvider) Pricing() (cloudprovider.PricingModel, errors.AutoscalerError) {
+	return nil, cloudprovider.ErrNotImplemented
```

Comment:
I am not exactly sure what `Pricing()` is used for in autoscaler, but the Hetzner API would make an implementation of this possible: https://docs.hetzner.cloud/#server-types

Reply:
On the other hand, this is not even implemented by Azure or AWS, so pretty save to leave it out...

Reply:
See https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/proposals/pricing.md

Reply:
Oh, that's pretty awesome actually, and would be a perfect fit here as the prices are indeed available through the API.

Reply:
Yes it would definitely be possible, we did not need this for our use case, so it was skipped. Feel free to implement, create a PR on our repo and I am happy to include.